creator-node/src/routes/users.js

```diff
@@ -41,16 +49,29 @@ module.exports = function (app) {
       return errorResponseBadRequest('Invalid data or signature')
     }
 
-    // signature data is valid and matches a user, proceed with verifying timestamp is
-    // within time boundary
     const theirTimestamp = parseInt(data.split(':')[1])
-    const ourTimestamp = Math.round((new Date()).getTime() / 1000)
-
-    if (Math.abs(theirTimestamp - ourTimestamp) > 3600) {
-      console.error(`Timestamp too old. User timestamp ${theirTimestamp}, Server timestamp ${ourTimestamp}`)
+    // Conditionally verify challenge based on request parameters
+    if (challenge) {
+      req.logger.info(`Detected challenge ${challenge}`)
+      const redisClient = req.app.get('redisClient')
+      const redisValue = await redisClient.get(address)
+      if (!redisValue) {
+        return errorResponseBadRequest('Missing challenge key')
+      }
+      const ourTimestamp = parseInt(redisValue)
+      if (theirTimestamp !== ourTimestamp) {
+        return errorResponseBadRequest(`Found invalid response: User timestamp ${theirTimestamp}, Server timestamp ${ourTimestamp}`)
+      } else {
+        req.logger.info(`Found equal signatures: User timestamp ${theirTimestamp}, Server timestamp ${ourTimestamp}`)
+      }
+    } else {
+      const ourTimestamp = Math.round((new Date()).getTime() / 1000)
+      if (Math.abs(theirTimestamp - ourTimestamp) > 3600) {
+        console.error(`Timestamp too old. User timestamp ${theirTimestamp}, Server timestamp ${ourTimestamp}`)
```

Comment:
Shouldn't this throw an error or return `errorResponseBadRequest` instead of just logging.
Also, shouldn't it be something like req.logger or import logger from utils w/ log level instead of console.error

Reply:
oh yes you right